shaders/src/light_indirect.fs

```diff
@@ -170,6 +204,38 @@ vec3 getReflectedVector(const PixelParams pixel, const vec3 n) {
     return getSpecularDominantDirection(n, r, pixel.roughness);
 }
 
+// This function returns an IBL lookup direction, taking into account the current IBL type (e.g. infinite spherical, finite/local sphere/box).
+vec3 GetAdjustedReflectedDirection(const MaterialInputs material, const PixelParams pixel, const vec3 normal) {
+    if (frameUniforms.iblTechnique == 0u) return getReflectedVector(pixel, normal);
+
+    // intersect the ray ray_pos + t * ray_dir with the finite geometry; done in the coordinate system of the finite geometry
+    vec3 ray_pos = getWorldPosition() + getWorldOffset() - frameUniforms.iblCenter;
+    vec3 ray_dir = normalize(reflect(-shading_view.xyz, normal.xyz));
+
+    vec3  r  = vec3(0.0); // resulting direction
+    float t0 = -1.0f;     // intersection parameter between ray and finite IBL geometry
+    
+    if (frameUniforms.iblTechnique == 1u) {
+        float R2 = frameUniforms.iblHalfExtents.r; // we store the squared radius to shave off a multiplication here
+        float A = 1.0; // in general, this should be dot(ray_dir, ray_dir) but we have just normalized it a couple of lines ago
+        float B = 2.0 * dot(ray_pos, ray_dir);
+        float C = dot(ray_pos, ray_pos) - R2;
+        vec2 roots = solveQuadratic(A, B, C);
+        t0 = getSmallestPositive(roots.x, roots.y);
+    }
+    else if (frameUniforms.iblTechnique == 2u) {
+        vec2 roots = intersectAABB(ray_pos, ray_dir, -frameUniforms.iblHalfExtents, frameUniforms.iblHalfExtents);
+
+        t0 = getSmallestPositive(roots.x, roots.y);
+    }
+
+    // translate results back to world space
+    vec3 intersection_point = ( t0 < 0.0 ) ? getReflectedVector(pixel, shading_normal) : frameUniforms.iblCenter + ray_pos + t0 * ray_dir;
+    r = normalize(intersection_point);
```

Comment:
Why do we need to add `frameUniforms.iblCenter`? Isn't the resulting reflection direction the direction from the IBL center to the intersection point?
On the figure this would mean `result = normalize(ibl_lookup_direction) = normalize(ibl_lookup_texel_world_space - ibl_center_world_space) = normalize( (ibl_center_world_space + ray_pos_ibl_local_space + t0 * ray_dir) - ibl_center_world_space ) = normalize(ray_pos_ibl_local_space + t0 * ray_dir)`. Is my reasoning flawed?
<img width="364" alt="Screenshot 2022-03-09 at 14 58 39" src="https://user-images.githubusercontent.com/91476779/157457274-9eb8ee95-e50c-4499-bd82-2869c3cf5aa5.png">